### PR TITLE
Fix and improve Regex doc

### PR DESCRIPTION
**Pull request by som-snytt**
As reported by @djvulee, there was an error in
the example explaining the confusing behavior
of `MatchIterator`. It's really confusing.

### CODE REVIEW

File: src/library/scala/util/matching/Regex.scala [hunk 1]
```diff
@@ -116,29 +118,41 @@ import java.util.regex.{ Pattern, Matcher }
...
+ *  with alternating calls to `hasNext` and `next`, `hasNext` has the additional
+ *  side effect of advancing the underlying matcher to the next unconsumed match.
+ *  This effect is visible in the `MatchData` representing the "current match".
  *
  *  {{{
- *  val num = """(\d+)""".r
- *  val all = num.findAllIn("123").toList  // List("123"), not List("123", "23", "3")
+ *  val r = "(ab+c)".r
+ *  val s = "xxxabcyyyabbczzz"
+ *  r.findAllIn(s).start    // 3
```
Comment (libratiger) [inline src/library/scala/util/matching/Regex.scala:132]: This is a little weird, it works under scala2.12.1, but output the following error under scala2.11.8


>scala> r.findAllIn(s).start
java.lang.IllegalStateException: No match available
  at java.util.regex.Matcher.start(Matcher.java:343)
  at scala.util.matching.Regex$MatchIterator.start(Regex.scala:761)
  ... 32 elided

**Comment by libratiger** [comment]
Looks better now, but there is a bug under 2.11.8 as I commented.

File: src/library/scala/util/matching/Regex.scala (hunk 1 above)
Comment (som-snytt) [reply src/library/scala/util/matching/Regex.scala:132]: Yes this is definitely 2.12 only. The fix was last summer.

BTW, I considered mentioning that everyone was bitten by the old behavior, but scaladoc doesn't usually do that. That's for release notes.

Comment (libratiger) [reply src/library/scala/util/matching/Regex.scala:132]: OK.

I am new to the Scala community, it looks good to me.
I will close my pull request.

File: src/library/scala/util/matching/Regex.scala [hunk 2]
```diff
@@ -116,29 +118,41 @@ import java.util.regex.{ Pattern, Matcher }
  *  while (mi.hasNext) {
  *    val d = mi.next
  *    if (mi.group(1).toInt < 1960) println(s"$d: An oldie but goodie.")
```
Comment (SethTisue) [inline src/library/scala/util/matching/Regex.scala:120]: a reference to SI- numbers, I'm sure

**Comment by SethTisue** [comment]
thanks Andrew and Djvu!

**Comment by libratiger** [comment]
Thanks for your merge! @SethTisue

**Pull request merged**